solidity/contracts/hooks/layer-zero/LayerZeroV2Hook.sol

```diff
@@ -0,0 +1,133 @@
+
+struct LayerZeroV2Metadata {
+    /// @dev the endpoint Id. prev dstChainId
+    uint32 eid;
+    /// @dev if the source transaction is cheaper than the amount of value passed, refund the additional amount to this address
+    address refundAddress;
+    /// @dev  parameters for the adapter service, e.g. send some dust native token to dstChain. prev adapterParam
+    bytes options;
+}
+
+contract LayerZeroV2Hook is AbstractMessageIdAuthHook, Indexed {
+    using StandardHookMetadata for bytes;
+    using Message for bytes;
+    using TypeCasts for bytes32;
+
+    ILayerZeroEndpointV2 public immutable lZEndpoint;
+
+    constructor(
+        address _mailbox,
+        uint32 _destinationDomain,
+        bytes32 _ism,
+        address _lZEndpoint
+    ) AbstractMessageIdAuthHook(_mailbox, _destinationDomain, _ism) {
+        lZEndpoint = ILayerZeroEndpointV2(_lZEndpoint);
+    }
+
+    // ============ External Functions ============
+
+    /// @inheritdoc AbstractMessageIdAuthHook
+    function _sendMessageId(
+        bytes calldata metadata,
+        bytes memory payload
+    ) internal override {
+        bytes calldata lZMetadata = metadata.getCustomMetadata();
+        (
+            uint32 eid,
+            address payable refundAddress,
+            bytes memory options
+        ) = parseLzMetadata(lZMetadata);
+
+        // Build and send message
+        MessagingParams memory msgParams = MessagingParams(
+            eid,
+            ism,
+            payload,
+            options,
+            false
+        );
+        lZEndpoint.send{value: msg.value}(msgParams, refundAddress);
+    }
+
+    /// @dev payInZRO is hardcoed to false because zro tokens should not be directly accepted
+    function _quoteDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal view virtual override returns (uint256) {
+        bytes calldata lZMetadata = metadata.getCustomMetadata();
+        (uint32 eid, , bytes memory options) = parseLzMetadata(lZMetadata);
+
+        // Build and quote message
+        MessagingParams memory msgParams = MessagingParams(
+            eid,
+            message.recipient(),
+            message.body(),
+            options,
+            false
+        );
+        MessagingFee memory msgFee = lZEndpoint.quote(
+            msgParams,
+            message.senderAddress()
+        );
+
+        return msgFee.nativeFee;
+    }
+
+    /**
+     * @notice Formats LayerZero metadata using default abi encoding
+     * @param layerZeroMetadata LayerZero specific metadata
+     * @return ABI encoded metadata
+     */
+    function formatLzMetadata(
+        LayerZeroV2Metadata calldata layerZeroMetadata
+    ) public pure returns (bytes memory) {
+        return
+            abi.encode(
+                layerZeroMetadata.eid,
+                layerZeroMetadata.refundAddress,
+                layerZeroMetadata.options
+            );
```

Comment:
ditto here
```suggestion
            abi.encode(layerZeroMetadata);
```
also in this case it should be possible to use packed because of the single dynamically sized member

Reply:
Ah good point. Updated